swarm/api/http/server.go

```diff
@@ -37,11 +37,33 @@ import (
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
 	"github.com/ethereum/go-ethereum/swarm/api"
 	"github.com/ethereum/go-ethereum/swarm/storage"
 	"github.com/rs/cors"
 )
 
+//setup metrics
+var (
+	postRawCount    = metrics.NewCounter("api.http.post.raw.count")
+	postRawFail     = metrics.NewCounter("api.http.post.raw.fail")
+	postFilesCount  = metrics.NewCounter("api.http.post.files.count")
+	postFilesFail   = metrics.NewCounter("api.http.post.files.fail")
+	deleteCount     = metrics.NewCounter("api.http.delete.count")
+	deleteFail      = metrics.NewCounter("api.http.delete.fail")
+	getCount        = metrics.NewCounter("api.http.get.count")
+	getFail         = metrics.NewCounter("api.http.get.fail")
+	getFileCount    = metrics.NewCounter("api.http.get.file.count")
+	getFileNotFound = metrics.NewCounter("api.http.get.file.notfound")
+	getFileFail     = metrics.NewCounter("api.http.get.file.fail")
+	getFilesCount   = metrics.NewCounter("api.http.get.files.count")
+	getFilesFail    = metrics.NewCounter("api.http.get.files.fail")
+	getListCount    = metrics.NewCounter("api.http.get.list.count")
+	getListFail     = metrics.NewCounter("api.http.get.list.fail")
```

Comment:
This is a lot of metrics, each of which cause overhead both in terms of memory and sending them to the metrics backend, are they all necessary? I don't think I'd ever consult `api.http.get.file.notfound` (I'd just want to know how many 4xx responses I'm sending, and then I'd consult my logs for more granular information).

Reply:
I agree partly with you. I believe our logs are actually very difficult to parse and to extract stats/intelligence from. This counters answer the question "How many GET /?list=true requests were made, how many failed, how many resulted in not found" quickly and easily. This isn't usually something you get from logs easily. Also, we decided to first focus on developers' needs for this first iteration of metrics. Thus, normal operation of a swarm node would be without the `--metrics` flag, and for development, it **may** well be worth collecting these stats, at least in a beginning. We could even think of a verbosity layer for stats, where the user can decide how many stats would be collected with a number (personally think to be overkill). Also, my assumption was that for *development*, the metrics overhead would be kind of limited (specifically for these counters, which just increment by one). Nevertheless agree to remove some (e.g. all "fail/notfound") if confirmed.

Reply:
I personally don't think we are measuring excessively. I doubt metrics have much overhead - we are literally incrementing a counter in-process and running a go-routine to flush the registry every 10sec. - this must be negligible compared to everything else in the process. If you believe we should measure and benchmark their overhead right now, fine, but I think this is an overkill.

Reply:
> I believe our logs are actually very difficult to parse and to extract stats/intelligence from. This counters answer the question "How many GET /?list=true requests were made, how many failed, how many resulted in not found" quickly and easily

So this is the real problem to solve, improve the logs. Structured, fine-grained, contextual logs can be parsed pretty easily with tools like grep and awk, and in a cluster can be sent through applications like [logstash](https://www.elastic.co/products/logstash) and then into [Elasticsearch](https://www.elastic.co/products/elasticsearch) which gives an easy way to answer the questions you are interested in, but also much more.

A single metric like `api.http.get.list.fail = 10` isn't all that useful by itself, if I see that then the first question I have is what is causing the list request to fail, which then leads me to consult the logs anyway.

If you think these metrics are useful _before_ we have better logging then I have no objection to leaving them in.

Reply:
If we have the `fail` metrics at least we will have a ballpark number and some intuition on the seriousness of the problem. I wouldn't worry too much if 0.1% of HTTP GET/ list=true requests fail for some reason.

Logging is ortogonal and we should address it as well, but it is out-of-scope for now.

Reply:
I agree with @nonsense but also agree that logs can be structured better, which would be a different issue though.